Sources/MySQL/Date.swift

```diff
@@ -83,34 +75,124 @@ extension Date {
-        } // TODO: in Linux
+        }
         
         // YYYY-MM-DD HH:MM:SS
         return EscapedQueryParameter( "'\(pad(num: comp.year ?? 0, digits: 4))-\(pad(num: comp.month ?? 0))-\(pad(num: comp.day ?? 0)) \(pad(num: comp.hour ?? 0)):\(pad(num: comp.minute ?? 0)):\(pad(num: comp.second ?? 0))'" )
     }
 }
+
+fileprivate func nanosecondsToString(_ nanosec: Int) -> String {
+    let nanosecSecond = Double(nanosec % 1_000_000_000)/1_000_000_000.0
+    var nanosecStr = String(format: "%.6f", nanosecSecond)
+    nanosecStr.removeFirst()
+    return String(nanosecStr)
+}
+
+extension DateComponents: QueryParameter {
+    public func queryParameter(option: QueryParameterOption) throws -> QueryParameterType {
+        if let year = self.year, let month = self.month, let day = self.day, let hour = self.hour, let minute = self.minute, let second = self.second {
+            var string = "'\(pad(num: year, digits: 4))-\(pad(num: month))-\(pad(num: day)) \(pad(num: hour)):\(pad(num: minute)):\(pad(num: second))"
+            if let nanosec = self.nanosecond {
+                string += nanosecondsToString(nanosec)
+            }
+            return EscapedQueryParameter(string + "'")
+        }
+        
+        if let hour = self.hour, let minute = self.minute, let second = self.second {
+            var string = "'\(pad(num: hour)):\(pad(num: minute)):\(pad(num: second))"
+            if let nanosec = self.nanosecond {
+                string += nanosecondsToString(nanosec)
+            }
+            return EscapedQueryParameter(string + "'")
+        }
+        if let year = self.year {
+            return EscapedQueryParameter("'\(pad(num: year))'")
+        }
+        throw QueryParameterError.dateComponentsError(self.description)
+    }
+}
+
+fileprivate let DateTimeRegex: NSRegularExpression = {
+    return try! NSRegularExpression(pattern: "^(\\d{4})-(\\d{2})-(\\d{2}) (\\d{2}):(\\d{2}):(\\d{2})\\.?(\\d*)$", options: [])
+}()
+
+fileprivate let TimeRegex: NSRegularExpression = {
+    return try! NSRegularExpression(pattern: "^(\\-?\\d{1,3}):(\\d{2}):(\\d{2})\\.?(\\d*)$", options: [])
+}()
+
+fileprivate let DateRegex: NSRegularExpression = {
+    return try! NSRegularExpression(pattern: "^(\\d{4})-(\\d{2})-(\\d{2})$", options: [])
+}()
+
+
+fileprivate func stringToNanoseconds<S: StringProtocol>(_ string: S) -> Int? {
+    guard string.count > 0 else {
+        return nil
+    }
+    guard let doubleValue = Double("0.\(string)") else {
+        return nil
+    }
+    return Int(doubleValue * 1_000_000_000.0)
+}
+
+extension DateComponents: SQLRawStringDecodable {
+    static func fromSQLValue(string: String) throws -> DateComponents {
+        if string.count == 4 {
+            // YEAR type
+            return DateComponents(year: Int(string))
+        }
+        let wholeRange = NSRange(string.startIndex..<string.endIndex, in: string)
+        if let match = DateTimeRegex.firstMatch(in: string, options: [], range: wholeRange) {
```

Comment:
A performance optimization would be to check the string length first. If the string does not have the required length, the expensive regex evaluation is not necessary.

Reply:
Thanks. I fixed for that.